File: Medium/2025-12-12-3433-Count-Mentions-Per-User/solution.java
import java.util.*;

public class Solution {
    public int[] countMentions(int numberOfUsers, String[][] events) {
        int[] mentions = new int[numberOfUsers];
        boolean[] online = new boolean[numberOfUsers];
        Arrays.fill(online, true);
        Map<Integer, Integer> offlineUntil = new HashMap<>();

        for (String[] event : events) {
            String type = event[0];
            int timestamp = Integer.parseInt(event[1]);

            if (type.equals("OFFLINE")) {
                int userId = Integer.parseInt(event[2]);
                online[userId] = false;
                offlineUntil.put(userId, timestamp + 60);
            } else if (type.equals("MESSAGE")) {
                String mentionsString = event[2];
                List<Integer> mentionedUsers = new ArrayList<>();

                if (mentionsString.equals("ALL")) {
                    for (int i = 0; i < numberOfUsers; i++) {
                        mentionedUsers.add(i);
                    }
                } else if (mentionsString.equals("HERE")) {
                    for (int i = 0; i < numberOfUsers; i++) {
                        if (online[i]) {
                            mentionedUsers.add(i);
                        }
                    }
                } else {
                    String[] ids = mentionsString.split(" ");
                    for (String id : ids) {
                        if (id.startsWith("id")) {
                            int userId = Integer.parseInt(id.substring(2));
                            mentionedUsers.add(userId);
                        }
                    }
                }

                for (int userId : mentionedUsers) {
                    if (online[userId]) {
                        mentions[userId] += Collections.frequency(Arrays.asList(mentionsString.split(" ")), "id" + userId);
                    }
                }
            }

            // Process offline users
            for (Map.Entry<Integer, Integer> entry : offlineUntil.entrySet()) {
                if (entry.getValue() <= timestamp) {
                    online[entry.getKey()] = true;
                }
            }
        }

        return mentions;
    }
}

File: Medium/2025-12-12-3433-Count-Mentions-Per-User/solution.py
class Solution:
    def countMentions(self, numberOfUsers: int, events: List[List[str]]) -> List[int]:
        mentions = [0] * numberOfUsers
        online = [True] * numberOfUsers
        offline_until = [-1] * numberOfUsers
        
        for event in events:
            event_type, timestamp, detail = event
            
            if event_type == "OFFLINE":
                user_id = int(detail)
                online[user_id] = False
                offline_until[user_id] = int(timestamp) + 60
            
            elif event_type == "MESSAGE":
                current_time = int(timestamp)
                for user_id in range(numberOfUsers):
                    if not online[user_id] and current_time < offline_until[user_id]:
                        continue
                
                if detail == "ALL":
                    for user_id in range(numberOfUsers):
                        mentions[user_id] += 1
                elif detail == "HERE":
                    for user_id in range(numberOfUsers):
                        if online[user_id]:
                            mentions[user_id] += 1
                else:
                    mentioned_users = detail.split()
                    for mention in mentioned_users:
                        if mention.startswith("id"):
                            user_id = int(mention[2:])
                            mentions[user_id] += 1
        
        return mentions